docs/reference/runtime-adapter.md

```diff
@@ -0,0 +1,251 @@
+//    ^^^^^^^^^^^^^^^
+```
+
+```ts twoslash [express]
+// @noErrors
+import type { Runtime, ExpressAdapter } from "@universal-middleware/core";
+
+export type Explain<A extends any> =
+  A extends Function
+    ? A
+    : {[K in keyof A]: A[K]} & unknown
+
+type RuntimeAdapter = Explain<ExpressAdapter>;
+
+// ---cut---
+import type { RuntimeAdapter } from "somelib/express";
+
+const runtime: RuntimeAdapter;
+
+// original express `req` and `res`. See https://expressjs.com/en/guide/using-middleware.html
+runtime.express;
+//      ^^^^^^^
+```
+
+```ts twoslash [fastify]
+// @noErrors
+import type { Runtime, FastifyAdapter } from "@universal-middleware/core";
+
+export type Explain<A extends any> =
+  A extends Function
+    ? A
+    : {[K in keyof A]: A[K]} & unknown
+
+type RuntimeAdapter = Explain<FastifyAdapter>;
+
+// ---cut---
+import type { RuntimeAdapter } from "somelib/fastify";
+
+const runtime: RuntimeAdapter;
+
+// original fastify `request` and `reply`. See https://fastify.dev/docs/latest/Guides/Getting-Started/
+runtime.fastify;
+//      ^^^^^^^
+```
+
+```ts twoslash [elysia]
+// @noErrors
+import type { Runtime, ElysiaAdapter } from "@universal-middleware/core";
+
+export type Explain<A extends any> =
+  A extends Function
+    ? A
+    : {[K in keyof A]: A[K]} & unknown
+
+type RuntimeAdapter = Explain<ElysiaAdapter>;
+
+// ---cut---
+import type { RuntimeAdapter } from "somelib/elysia";
+
+const runtime: RuntimeAdapter;
+
+// original elysia context. See https://elysiajs.com/essential/handler#context
+runtime.elysia;
+//      ^^^^^^
+```
+
+```ts twoslash [vercel-node]
+// @noErrors
+import type { Runtime, VercelNodeAdapter } from "@universal-middleware/core";
+
+export type Explain<A extends any> =
+  A extends Function
+    ? A
+    : {[K in keyof A]: A[K]} & unknown
+
+type RuntimeAdapter = Explain<VercelNodeAdapter>;
+
+// ---cut---
+import type { RuntimeAdapter } from "somelib/vercel-node";
+
+const runtime: RuntimeAdapter;
+
+// original node `req` and `res`. See https://expressjs.com/en/guide/using-middleware.html
+const originalParameters = runtime["vercel-node"];
+//    ^^^^^^^^^^^^^^^^^^
+```
+
+:::
+
+### Common properties
+
+All `runtime` objects can also contain the following properties:
+
+```ts twoslash
+// @noErrors
+import type { Runtimes, Adapters } from "@universal-middleware/core";
+import type { IncomingMessage, ServerResponse } from "node:http";
+
+export type Explain<A extends any> =
+  A extends Function
+    ? A
+    : { [K in keyof A]: A[K] } & unknown
+
+interface _RuntimeAdapter {
+  adapter: Adapters;
+  runtime: Runtimes;
+  params?: Record<string, string>;
+  req?: IncomingMessage;
+  res?: ServerResponse;
+}
+
+type RuntimeAdapter = Explain<_RuntimeAdapter>;
+
+// ---cut---
+const runtime: RuntimeAdapter;
+
+// identify which adapter is currently used
+runtime.adapter;
+//      ^^^^^^^
+// runtime environment identifier. See https://runtime-keys.proposal.wintercg.org/
```

Comment:
The runtime name ('hono', 'express', ...). Follows https://runtime-keys.proposal.wintercg.org/